crates/core/src/nucleus/state.rs

```diff
@@ -214,13 +220,73 @@ impl AddressableContent for NucleusStateSnapshot {
     }
 }
 
+#[derive(Clone, Default, Debug, PartialEq, Serialize)]
+pub struct ZomeFnCallState {
+    hdk_fn_invocations: Vec<(HdkFnCall, Option<HdkFnCallResult>)>,
+}
+
+impl ZomeFnCallState {
+    pub fn begin_hdk_call(&mut self, call: HdkFnCall) {
+        self.hdk_fn_invocations.push((call, None))
+    }
+
+    pub fn end_hdk_call(
+        &mut self,
+        call: HdkFnCall,
+        result: HdkFnCallResult,
+    ) -> Result<(), HolochainError> {
+        if let Some((current_call, current_result)) = self.hdk_fn_invocations.pop() {
+            if call != current_call {
+                Err(HolochainError::new(
+                    "HDK call other than the current call was ended.",
+                ))
+            } else if current_result.is_some() {
+                Err(HolochainError::new(
+                    "Ending and HDK which was already ended.",
```

Comment:
Is this what was meant?

```suggestion
                    "Ending an HDK call which was already ended.",
```